platform/search/src/GenericSearchProvider.js

```diff
@@ -0,0 +1,267 @@
+                var ids = [],
+                    id;
+                
+                // If we have the results from a search 
+                if (event.data.request === 'search') {
+                    // Convert the ids given from the web worker into domain objects
+                    for (id in event.data.results) {
+                        ids.push(id);
+                    }
+                    objectService.getObjects(ids).then(function (objects) {
+                        var searchResults = [],
+                            id;
+                        
+                        // Create searchResult objects
+                        for (id in objects) {
+                            searchResults.push({
+                                object: objects[id],
+                                id: id,
+                                score: event.data.results[id]
+                            });
+                        }
+                        
+                        // Resove the promise corresponding to this 
+                        pendingQueries[event.data.timestamp].resolve({
+                            hits: searchResults,
+                            total: event.data.total,
+                            timedOut: event.data.timedOut
+                        });
+                    });
+                }
+            }
+            
+            worker.onmessage = handleResponse;
+            
+            // Helper function for getItems(). Indexes the tree.
+            function indexItems(nodes) {
+                nodes.forEach(function (node) {
+                    var id = node && node.getId && node.getId();
+                    
+                    // If we have not yet indexed this item, index it 
+                    if (!indexed[id]) {
+                        // Index each item with the web worker
+                        indexItem(node);
+                        indexed[id] = true;
+                        
+                        if (node.hasCapability && node.hasCapability('composition')) {
+                            // This node has children
+                            
+                            // Make sure that this is async, so doesn't block up page
+                            $timeout(function () {
+                                // Get the children... 
+                                node.useCapability('composition').then(function (children) {
+                                    $timeout(function () {
+                                        // ... then index the children
+                                        if (children.constructor === Array) {
+                                            indexItems(children);
+                                        } else {
+                                            indexItems([children]);
+                                        }
+                                    }, 100);
+                                });
+                            }, 100);
```

Comment:
Under 10k items?  That should be trivial in terms of performance-- there's something else here for sure.

As you mentioned, Line 142 is the performance culprit. Unfortunately, line 142 contains two different calls: useCapability, and your callback function that executes when useCapability completes.

Can you isolate one or the other as the performance issue?  If the problem is with useCapability, then the performance issue is outside of the scope of your code and you can very safely ignore it.

Regardless, the recursion issue with mutation listeners should be avoided-- the change I recommended for skipping indexed items would handle that for you.